### DISCUSSION TITLE
質問の検索機能についてお聞かせください！

### DISCUSSION THREAD

**Discussion by tanaka-asami415** [increments/qiita-discussions#450] [要望・アイデア・不具合]
https://github.com/increments/qiita-discussions/discussions/239
https://github.com/increments/qiita-discussions/discussions/201
上記のDiscussionでもご意見をいただいた、質問の検索機能について
他にも「質問を検索できるようにして欲しい」というご意見は多くいただいています。

- 過去に同様の質問がないを調べることができず、同じ質問が繰り返される
- せっかくの質問と回答が活用されにくい（知見が活用されていない）

我々としても、検索できないことで起きている上記のような問題を解決したいと考え、質問を検索できるようにする予定です！

そこで、どんな条件で絞り込みができるといいか、みなさまにご意見をお聞かせいただきたいです！🙏

## どのように検索できるようにするか
以下UIを想定しております

- 現在ヘッダーにある検索機能で、質問ページで検索した際には質問の検索ができる
- 検索した後は、記事の検索結果と同様に、「オプションを追加」や「並び順の変更」「表示の切替（記事の検索結果だと「すべて」と「ストック済み」といったもの）」ができる

**参考画像**

<details><summary>検索場所</summary>

![画像①](https://user-images.githubusercontent.com/103916185/225203075-eaea5f84-a25f-4f68-8022-7d31792286f9.png)
</details>

<details><summary>オプションを追加</summary>

![検索オプション](https://user-images.githubusercontent.com/103916185/225203124-8d649327-5a0f-43ac-aa22-567121ca94e1.png)
</details>

<details><summary>並び順の変更と表示の切替</summary>

![並び順の変更と表示の切替](https://user-images.githubusercontent.com/103916185/225203639-bd719d87-339d-439d-865d-66b25381172a.png)
</details>

#### 1. どのような項目で検索を絞り込みたいですか？

（例）
  - タイトル、ユーザー名で絞り込みたい
  - 未回答 or 回答済みの質問の表示を切り替えたい
  - 所属しているOrganizationで絞り込みたい
  - 回答数で並び替えたい

#### 2. 上記の抽出条件は、主にどんな時に必要かを教えてください
（例）
- 回答（質問）する際に必要 / 閲覧する際に必要など…

#### 3. その他
- 何かあればご記載ください

---

質問機能について、以下を[Qiitaの開発ロードマップ](https://github.com/orgs/increments/projects/8/views/1)に掲載いたしました！✨
・回答に対するコメントでMarkdownが書けるようにする
・質問者・回答者以外も回答に対してコメントができるようにする

Qiitaの開発ロードマップについては https://github.com/increments/qiita-discussions/discussions/449 をご確認ください。

**Comment by radian-jp** [comment]
1. どのような項目で検索を絞り込みたいですか？

通常の記事の検索と同じ仕様で十分だと思われますが、検索時に利用できるオプションを確認するまでの方法をシンプルにするべきだと思います。

下記は、[スタック・オーバーフロー](https://ja.stackoverflow.com/)の検索ワード入力箇所ですが、フォーカスを移すと、基本的な検索方法と詳細な検索ヘルプに直接飛ぶリンクがポップアップ表示されます。
![search](https://user-images.githubusercontent.com/12453432/225208146-2ec1126f-8a2d-4c36-bc74-b8e69bfbc34a.png)

2. 上記の抽出条件は、主にどんな時に必要かを教えてください

3. その他

そもそも質問する人は検索をしない人が多いように見受けられるので、実際に投稿される前に、タイトルとタグで類似質問・記事を抽出して一覧表示する画面をワンクッション挟むとよいのかもしれません。

**Reply by radian-jp** [reply]
あと、qiita-discussions全体を見ている人の割合はそれほど多くないのではないかと思われますので、もし広く利用者に意見を求めるのであれば、QiitaまたはQiita質問の所からもリンクを貼った方が良いのではないかと思います。（私もたまたま目に入ったからクリックしただけです）

**Reply by tanaka-asami415** [reply]
@radian-jp 
ありがとうございます！🙏
Discussionでは、活発にご意見をくださる方が見てくださっていることもあり、優先して実施しています！
Discussionにプラスして、ユーザーインタビューも実施しているのですが、
今後の展開によっては、ご提案いただいた方法も検討してみます！✨

**Comment by yoshi389111** [comment]
最近はあまり回答もしていないので、以前思ったことですが（主に回答者としての意見です）

* 回答数で絞り込みたいです。例えば「回答数0件のもの」とか「回答数が2件以下のもの」とか
* 「クローズ」済/未済で絞り込みたいです。
* 「Q&A」あるいは「意見交換」の種類で絞り込みたいです。

理由はどれも同じで、回答がされていないで埋もれている質問があると、質問者がかわいそうなので、埋もれている未回答のものや、すでに回答はあるけどもう少し補足できるものがないかを見られると、回答しやすいからです。

* 回答者での絞り込みができるとうれしいです。

おそらくユーザでの絞り込み(例での `user:Qiita` ）は、質問者での絞り込みだと思うのですが、回答者でも絞り込みがしたいです。
これは、例えば同じような質問があって、以前自分が回答したｘｘｘｘについての質問・回答のへのリンクを張りたいときに、検索しやすくするためです。
今でも、自分が回答したものは、ユーザページから見れますが、あちらは絞り込みができないですよね。

* 記事/質問を横断した検索ができるとよいかも？

これはあまり積極的に必要な理由はなく、例えばエラーの対策を検索するときに、両方一緒に検索出来た方がよいかな？と思った程度です。
大抵は外部の検索サイト(googleとか)を使うから不要かもしれないですが。

**Comment by blue32a** [comment]
# 1. どのような項目で検索を絞り込みたいですか？

回答者としては次の検索ができればと思います。

- 新着
- 技術領域（タグ）
- 購読済み

質問者としては、検索機能よりはSEO対策になると思います。問題解決の方法は特定サービスに限定するものではないからです。Google検索で横断的に検索されることが多いのではないでしょうか。
またすでにアイデアとして出ていますが、質問の投稿時に類似の質問や記事を表示するのは良いと思います。

# 2. 上記の抽出条件は、主にどんな時に必要かを教えてください

新着順が必要なのは次の２点です。

1. 現在の投稿数だと１日１度新着順を見れば、新しいものは確認できるため
2. なるべく新しい質問に回答したいため

新しい質問を優先したい理由としては、「古い質問はすでに状況が変わっているのではないか」と思うからです。
まだ回答が求められているのか不明のため、頑張って未回答の質問を検索するよりも、新着を優先しています。

技術領域（タグ）は自分が回答できそうな範囲で絞り込みをしたいと思いますが、タグが正確ではないケースもあるので優先度は低いです。

購読は「回答できないが興味がある」というケースに使用しているので、後から回答を確認するために購読済みを検索できたらと思います。

**Comment by tanaka-asami415** [comment]
@radian-jp @yoshi389111 @blue32a 
みなさんご意見をいただきありがとうございます！🙇

頂いた内容を元に、実際の仕様と照らし合わせながら実装計画を進めていきます。
具体的な内容が決まりましたら、改めてご報告させていただきます！

**Comment by tanaka-asami415** [comment]
ご協力いただきありがとうございます！
昨日ついにベータ版をリリースすることができました！🙇✨

- [リリースノート](https://qiita.com/release-notes/2023-07-18)
- [ヘルプ](https://help.qiita.com/ja/articles/qiita-search-options)

### 検索場所ついて
当初は、「現在ヘッダーにある検索機能で、質問ページで検索した際には質問の検索ができる」（検索結果に質問を表示）という想定でおりましたが、
検索結果ページで「質問」に切り替えていただくUIとなっております。

### 検索オプションについて
記事で利用できるオプション（一部利用できないものもあり） ＋ 下記の絞り込みが可能なオプションをご使用いただけます。
詳細はヘルプをご確認ください。

- 質問者、回答者
- 回答数
- フォーマット（「Q&A」or「意見交換」）
- クローズ済 or 回答募集中

### 並び順の変更と表示の切替について
検索後に「新着順」等の並び替え、「購読済み」の質問に絞ることが可能です。

@yoshi389111さん @blue32aさん
概ねご記載いただいたリクエストは網羅できたのではないかと思います🙏

@blue32aさん
検索結果に対して新着順に表示されたい場合は、並べ替えをご利用いただき
全質問の新着はフィードで確認することが可能です。
https://help.qiita.com/ja/articles/qiita-question-feed

@radian-jpさん
投稿前のサジェスト等については、今後の改善で検討していきたいと思います🙏

みなさんご協力いただきありがとうございました！
